src/condor_utils/file_transfer.cpp

```diff
@@ -1855,6 +1855,30 @@ FileTransfer::callClientCallback()
 	}
 }
 
+bool
+FileTransfer::PipeReadFullString(std::string& buf, const int nBytes) {
+	int nleft = nBytes;
+
+	while (nleft > 0) {
+		char* tmp_buf = new char[nleft+1];
+		ASSERT(tmp_buf);
+
+		int nread = daemonCore->Read_Pipe(TransferPipe[0],
+		                                  tmp_buf,
+                                                  nleft);
+
+		tmp_buf[nread] = '\0';
+		nleft -= nread;
+		buf += tmp_buf;
```

Comment:
While adding a NUL to the end protects against a buffer overflow, there's nothing here preventing a buffer _under_flow; it's perfectly legal to read NUL bytes from a pipe.

You can do use `insert()` to copy a span from a buffer into an offset in a `std::string`.